jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java

```diff
@@ -927,6 +928,13 @@ default int size()
         return size;
     }
 
+    static Map<String, List<String>> asMap(HttpFields fields)
+    {
+        Map<String, List<String>> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+        fields.getFieldNamesCollection().forEach(name -> headers.putIfAbsent(name, fields.getValuesList(name)));
+        return headers;
+    }
```

Comment:
This could be more efficiently implemented without the temporary name collection:
```suggestion
    static Map<String, List<String>> asMap(HttpFields fields)
    {
        Map<String, List<String>> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
        for (HttpField f : fields)
        {
            if (!headers.containsKey(f.getName()))
            {
                HttpHeader header = f.getHeader();
                headers.put(f.getName(), header == null ? fields.getValuesList(f.getName()) : fields.getValuesList(header));
            }
        }
        return headers;
    }
```

Reply:
Wouldn't it be easier to read as ...

``` java
    static Map<String, List<String>> asMap(HttpFields fields)
    {
        Map<String, List<String>> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
        for (HttpField f : fields)
            headers.putIfAbsent(f.getName(), fields.getValueList(f.getHeader());
        return headers;
    }
```

Reply:
@joakime the problem with that version, is that if a header is repeated, then the value list is generated multiple times, but only used once.

Reply:
@gregw then use `computeIfAbsent` via a lamda ...

``` java
    static Map<String, List<String>> asMap(HttpFields fields)
    {
        Map<String, List<String>> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
        for (HttpField f : fields)
            headers.computeIfAbsent(f.getName(), k -> fields.getValueList(f.getHeader());
        return headers;
    }
```

Reply:
Resolving this as now we are wrapping the `HttpFields` instead of creating a `TreeMap`.